presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java

```diff
@@ -399,34 +405,59 @@ public boolean trySetMemoryReservation(long newMemoryReservation)
 
     public synchronized boolean isMemoryRevokingRequested()
     {
-        return memoryRevokingRequestedFuture.isDone();
+        return memoryRevokingRequested;
     }
 
     /**
      * Returns how much revocable memory will be revoked by the operator
      */
-    public synchronized long requestMemoryRevoking()
-    {
-        boolean alreadyRequested = isMemoryRevokingRequested();
-        if (!alreadyRequested && revocableMemoryReservation > 0) {
-            memoryRevokingRequestedFuture.set(null);
-            return revocableMemoryReservation;
+    public long requestMemoryRevoking()
+    {
+        long revocableMemory = 0L;
+        Runnable listener = null;
+        synchronized (this) {
+            if (!isMemoryRevokingRequested() && revocableMemoryReservation > 0) {
+                memoryRevokingRequested = true;
+                revocableMemory = revocableMemoryReservation;
+                listener = memoryRevocationRequestListener;
+            }
+        }
+        if (listener != null) {
+            runListener(listener);
         }
-        return 0;
+        return revocableMemory;
     }
 
     public synchronized void resetMemoryRevokingRequested()
     {
-        SettableFuture<?> currentFuture = memoryRevokingRequestedFuture;
-        if (!currentFuture.isDone()) {
-            return;
+        memoryRevokingRequested = false;
+    }
+
+    public void setMemoryRevocationRequestListener(Runnable listener)
+    {
+        requireNonNull(listener, "listener is null");
+
+        boolean shouldNotify;
+        synchronized (this) {
+            checkState(memoryRevocationRequestListener == null, "listener already set");
+            memoryRevocationRequestListener = listener;
+            shouldNotify = memoryRevokingRequested;
+        }
+        // if memory revoking is requested immediately run the listener
+        if (shouldNotify) {
+            runListener(listener);
         }
-        memoryRevokingRequestedFuture = SettableFuture.create();
     }
 
-    public synchronized SettableFuture<?> getMemoryRevokingRequestedFuture()
+    private static void runListener(Runnable listener)
     {
-        return memoryRevokingRequestedFuture;
+        requireNonNull(listener, "listener is null");
+        try {
+            listener.run();
+        }
+        catch (RuntimeException e) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "Exception while running the listener", e);
```

Comment:
Sad thing is that when this exc is propagated up to `MemoryRevokingScheduler`, it doesn't seem to fail a query. :/

No change requested here.